File: Engine/Ios/Sources/Combs/CombsEngine.swift
import Foundation

/// CombsEngine — Swift wrapper over the combs native core (libcombs_ffi).
///
/// Thin glue over the stable C ABI (combs.h); all inference lives in the
/// Rust core (Metal via wgpu).
///
/// ```swift
/// let engine = try CombsEngine(configJson: #"{"model_dir": ".../smollm2-135m"}"#)
/// try engine.chatCompletion(#"{"messages":[{"role":"user","content":"hi"}]}"#,
///                           requestId: "req-1") { event in
///     print(event) // delta / done / error JSON
/// }
/// ```
public final class CombsEngine {
    public typealias StreamHandler = (String) -> Void

    private let handle: OpaquePointer

    /// Unmanaged callback context passed through the C ABI.
    private class CallbackBox {
        let handler: StreamHandler
        init(_ handler: @escaping StreamHandler) { self.handler = handler }
    }

    /// Device capabilities JSON (buffer limits, backend, features).
    public static var deviceCaps: String {
        guard let caps = combs_device_caps_json() else { return "{}" }
        defer { combs_string_free(caps) }
        return String(cString: caps)
    }

    /// Last error on this thread.
    public static var lastError: String {
        guard let err = combs_last_error() else { return "" }
        return String(cString: err)
    }

    public init(configJson: String) throws {
        var handle: OpaquePointer?
        configJson.withCString { ptr in
            handle = combs_engine_create(ptr)
        }
        guard let handle else {
            throw NSError(
                domain: "CombsEngine",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "engine creation failed: \(CombsEngine.lastError)"]
            )
        }
        self.handle = handle
    }

    deinit {
        combs_engine_destroy(handle)
    }

    /// Model metadata JSON (architecture, vocab, context, eos ids).
    public var metadata: String {
        guard let md = combs_engine_metadata_json(handle) else { return "{}" }
        defer { combs_string_free(md) }
        return String(cString: md)
    }

    /// Runs a chat completion (BLOCKS — call from a background queue).
    /// Streams JSON events to `handler`.
    public func chatCompletion(
        _ requestJson: String,
        requestId: String,
        handler: @escaping StreamHandler
    ) throws {
        let box = CallbackBox(handler)
        let userData = Unmanaged.passRetained(box).toOpaque()
        defer { Unmanaged<CallbackBox>.fromOpaque(userData).release() }

        let callback: CombsStreamCallback = { eventJson, userData in
            guard let eventJson, let userData else { return }
            let box = Unmanaged<CallbackBox>.fromOpaque(userData).takeUnretainedValue()
            box.handler(String(cString: eventJson))
        }

        let rc = requestJson.withCString { reqPtr in
            requestId.withCString { idPtr in
                combs_chat_completion(handle, reqPtr, idPtr, callback, userData)
            }
        }
        guard rc == 0 else {
            throw NSError(
                domain: "CombsEngine",
                code: Int(rc),
                userInfo: [NSLocalizedDescriptionKey: "chat completion failed: \(CombsEngine.lastError)"]
            )
        }
    }

    /// Requests cancellation of an in-flight completion.
    public func cancel(requestId: String) {
        requestId.withCString { ptr in
            _ = combs_cancel(ptr)
        }
    }
}

File: Engine/Android/combs-android/src/main/kotlin/ai/combs/CombsEngine.kt
package ai.combs

/**
 * CombsEngine — Kotlin wrapper over the combs native core (libcombs_ffi.so).
 *
 * Mirrors the LiteRT-LM Engine/Conversation shape: create with a JSON
 * config, stream chat completions, cancel by request id.
 *
 * ```kotlin
 * val engine = CombsEngine.create("""{"model_dir": "/data/models/smollm2"}""")
 * engine.chatCompletion("""{"messages":[{"role":"user","content":"hi"}]}""") { event ->
 *     // {"type":"delta","text":"..."} / {"type":"done",...} / {"type":"error",...}
 * }
 * engine.close()
 * ```
 */
class CombsEngine private constructor(private val handle: Long) : AutoCloseable {

    fun interface StreamCallback {
        fun onEvent(eventJson: String)
    }

    companion object {
        init {
            System.loadLibrary("combs_jni")
            System.loadLibrary("combs_ffi")
        }

        /** Device capabilities JSON (buffer limits, backend, features). */
        @JvmStatic
        external fun nativeDeviceCaps(): String?

        @JvmStatic
        private external fun nativeCreate(configJson: String): Long

        @JvmStatic
        private external fun nativeDestroy(handle: Long)

        @JvmStatic
        private external fun nativeMetadata(handle: Long): String?

        @JvmStatic
        private external fun nativeChatCompletion(
            handle: Long,
            requestJson: String,
            requestId: String,
            callback: StreamCallback,
        ): Int

        @JvmStatic
        private external fun nativeCancel(requestId: String): Int

        @JvmStatic
        external fun nativeLastError(): String?

        /**
         * Creates an engine. `configJson`:
         * `{"model_dir": "...", "max_seq_len": 4096, "kv_cache": "paged"}`
         */
        fun create(configJson: String): CombsEngine {
            val handle = nativeCreate(configJson)
            check(handle != 0L) { "combs engine creation failed: ${nativeLastError()}" }
            return CombsEngine(handle)
        }
    }

    /** Model metadata JSON (architecture, vocab, context, eos ids). */
    fun metadata(): String = nativeMetadata(handle) ?: "{}"

    /**
     * Runs a chat completion (BLOCKS the calling thread — call from a
     * background executor). Streams JSON events to `callback`.
     */
    fun chatCompletion(requestJson: String, requestId: String, callback: StreamCallback) {
        val rc = nativeChatCompletion(handle, requestJson, requestId, callback)
        check(rc == 0) { "chat completion failed: ${nativeLastError()}" }
    }

    /** Requests cancellation of an in-flight completion. */
    fun cancel(requestId: String) {
        nativeCancel(requestId)
    }

    override fun close() {
        nativeDestroy(handle)
    }
}
